File: cpp/Ice/throughput/ThroughputI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2015 ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <Ice/Ice.h>
#include <ThroughputI.h>

ThroughputI::ThroughputI() :
    _byteSeq(Demo::ByteSeqSize),
    _stringSeq(Demo::StringSeqSize, "hello"),
    _stringViewSeq(Demo::StringSeqSize, "hello"),
    _structSeq(Demo::StringDoubleSeqSize),
    _fixedSeq(Demo::FixedSeqSize),
    _warmup(false)
{
    int i;
    for(i = 0; i < Demo::StringDoubleSeqSize; ++i)
    {
        _structSeq[i].s = "hello";
        _structSeq[i].d = 3.14;
    }
    for(i = 0; i < Demo::FixedSeqSize; ++i)
    {
        _fixedSeq[i].i = 0;
        _fixedSeq[i].j = 0;
        _fixedSeq[i].d = 0;
    }
}

bool
ThroughputI::needsWarmup(const Ice::Current&)
{
    _warmup = false;
    return false;
}

void
ThroughputI::startWarmup(const Ice::Current&)
{
    _warmup = true;
}

void
ThroughputI::endWarmup(const Ice::Current&)
{
    _warmup = false;
}

void
ThroughputI::sendByteSeq(const std::pair<const Ice::Byte*, const Ice::Byte*>&, const Ice::Current&)
{
}

void
ThroughputI::recvByteSeq_async(const Demo::AMD_Throughput_recvByteSeqPtr& cb, const Ice::Current&)
{
    std::pair<const Ice::Byte*, const Ice::Byte*> ret;
    if(_warmup)
    {
        Demo::ByteSeq empty(1);
        ret.first = &empty[0];
        ret.second = ret.first + empty.size();
    }
    else
    {
        ret.first = &_byteSeq[0];
        ret.second = ret.first + _byteSeq.size();
    }
    cb->ice_response(ret);
}

void
ThroughputI::echoByteSeq_async(const Demo::AMD_Throughput_echoByteSeqPtr& cb, 
                         const std::pair<const Ice::Byte*, const Ice::Byte*>& seq, const Ice::Current&)
{
    cb->ice_response(seq);
}

void
ThroughputI::sendStringSeq(const std::vector<Util::string_view>&, const Ice::Current&)
{
}

void
ThroughputI::recvStringSeq_async(const Demo::AMD_Throughput_recvStringSeqPtr& cb, 
                                 const Ice::Current&)
{
    if(_warmup)
    {
        cb->ice_response(std::vector<Util::string_view>(0));
    }
    else
    {
        cb->ice_response(_stringViewSeq);
    }
}

void
ThroughputI::echoStringSeq_async(const Demo::AMD_Throughput_echoStringSeqPtr& cb, 
                                 const std::vector<Util::string_view>& seq, const Ice::Current&)
{
    cb->ice_response(seq);
}

void
ThroughputI::sendStructSeq(const Demo::StringDoubleSeq&, const Ice::Current&)
{
}

Demo::StringDoubleSeq
ThroughputI::recvStructSeq(const Ice::Current&)
{
    if(_warmup)
    {
        return Demo::StringDoubleSeq();
    }
    else
    {
        return _structSeq;
    }
}

Demo::StringDoubleSeq
ThroughputI::echoStructSeq(const Demo::StringDoubleSeq& seq, const Ice::Current&)
{
    return seq;
}

void
ThroughputI::sendFixedSeq(const Demo::FixedSeq&, const Ice::Current&)
{
}

Demo::FixedSeq
ThroughputI::recvFixedSeq(const Ice::Current&)
{
    if(_warmup)
    {
        return Demo::FixedSeq();
    }
    else
    {
        return _fixedSeq;
    }
}

Demo::FixedSeq
ThroughputI::echoFixedSeq(const Demo::FixedSeq& seq, const Ice::Current&)
{
    return seq;
}

void
ThroughputI::shutdown(const Ice::Current& c)
{
    c.adapter->getCommunicator()->shutdown();
}

File: java-compat/Ice/throughput/ThroughputI.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

import Demo.*;

public final class ThroughputI extends _ThroughputDisp
{
    public
    ThroughputI()
    {
        _byteSeq = new byte[ByteSeqSize.value];

        _stringSeq = new String[StringSeqSize.value];
        for(int i = 0; i < StringSeqSize.value; ++i)
        {
            _stringSeq[i] = "hello";
        }

        _structSeq = new StringDouble[StringDoubleSeqSize.value];
        for(int i = 0; i < StringDoubleSeqSize.value; ++i)
        {
            _structSeq[i] = new StringDouble();
            _structSeq[i].s = "hello";
            _structSeq[i].d = 3.14;
        }

        _fixedSeq = new Fixed[FixedSeqSize.value];
        for(int i = 0; i < FixedSeqSize.value; ++i)
        {
            _fixedSeq[i] = new Fixed();
            _fixedSeq[i].i = 0;
            _fixedSeq[i].j = 0;
            _fixedSeq[i].d = 0;
        }
    }

    @Override
    public boolean
    needsWarmup(Ice.Current current)
    {
        _warmup = false;
        return _needsWarmup;
    }

    @Override
    public void
    startWarmup(Ice.Current current)
    {
        _warmup = true;
    }

    @Override
    public void
    endWarmup(Ice.Current current)
    {
        _warmup = false;
        _needsWarmup = false;
    }

    @Override
    public void
    sendByteSeq(byte[] seq, Ice.Current current)
    {
    }

    @Override
    public byte[]
    recvByteSeq(Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupByteSeq;
        }
        else
        {
            return _byteSeq;
        }
    }

    @Override
    public byte[]
    echoByteSeq(byte[] seq, Ice.Current current)
    {
        return seq;
    }

    @Override
    public void
    sendStringSeq(String[] seq, Ice.Current current)
    {
    }

    @Override
    public String[]
    recvStringSeq(Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupStringSeq;
        }
        else
        {
            return _stringSeq;
        }
    }

    @Override
    public String[]
    echoStringSeq(String[] seq, Ice.Current current)
    {
        return seq;
    }

    @Override
    public void
    sendStructSeq(StringDouble[] seq, Ice.Current current)
    {
    }

    @Override
    public StringDouble[]
    recvStructSeq(Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupStructSeq;
        }
        else
        {
            return _structSeq;
        }
    }

    @Override
    public StringDouble[]
    echoStructSeq(StringDouble[] seq, Ice.Current current)
    {
        return seq;
    }

    @Override
    public void
    sendFixedSeq(Fixed[] seq, Ice.Current current)
    {
    }

    @Override
    public Fixed[]
    recvFixedSeq(Ice.Current current)
    {
        if(_warmup)
        {
            return _warmupFixedSeq;
        }
        else
        {
            return _fixedSeq;
        }
    }

    @Override
    public Fixed[]
    echoFixedSeq(Fixed[] seq, Ice.Current current)
    {
        return seq;
    }

    @Override
    public void
    shutdown(Ice.Current current)
    {
        current.adapter.getCommunicator().shutdown();
    }

    private byte[] _byteSeq;
    private String[] _stringSeq;
    private StringDouble[] _structSeq;
    private Fixed[] _fixedSeq;

    private byte[] _warmupByteSeq = new byte[1];
    private String[] _warmupStringSeq = new String[1];
    private StringDouble[] _warmupStructSeq = new StringDouble[] { new StringDouble() };
    private Fixed[] _warmupFixedSeq = new Fixed[] { new Fixed() };

    private boolean _needsWarmup = true;
    private boolean _warmup = false;
}
